src/Lecture4.hs

```diff
@@ -257,7 +384,9 @@ the file doesn't have any products.
 -}
 
 calculateStats :: String -> String
-calculateStats = error "TODO"
+calculateStats s = case (mapMaybe parseRow . lines) s of
+  [] -> "This file does not contain any product."
+  rows -> (displayStats . combineRows . fromList) rows
```

Comment:
Here you can use the `:|` constructor directly and pattern match on the list to avoid using the partial function `fromList` and guarantee the creation of a NonEmpty list.
```suggestion
  row : rows -> displayStats $ combineRows $ row :| rows
```